File: src/shared/executor.ts
import {
  MAX_LOGS,
  errorMessage,
  logText,
  sanitize,
  type ExecutionLog,
  type Outcome,
  type WireRequest,
  type ProgressLog,
} from "./protocol";
import { scheduleTimeout } from "./timer";
export interface Runtime {
  resource: string;
  require?: unknown;
  exports: unknown;
  print: (level: ExecutionLog["level"], message: string) => void;
  game?: <T>(callback: () => T) => Promise<T>;
  progress?: (packet: ProgressLog) => void;
}

interface Context {
  id: string;
  resource: string;
  alive(): boolean;
  log(...values: unknown[]): void;
  sleep(ms: number): Promise<void>;
  game<T>(callback: () => T): Promise<T>;
  onCleanup(callback: () => void): void;
}

type Snippet = (
  ctx: Context,
  console: Record<string, (...args: unknown[]) => void>,
  require: unknown,
  exports: unknown,
) => Promise<unknown>;
const AsyncFunction = Object.getPrototypeOf(async function () {})
  .constructor as new (...args: string[]) => Snippet;

export class JavascriptExecutor {
  private readonly active = new Map<string, () => void>();
  constructor(private readonly runtime: Runtime) {}

  cancel(id: string): void {
    this.active.get(id)?.();
  }
  close(): void {
    for (const cancel of [...this.active.values()]) cancel();
  }

  async execute(request: WireRequest): Promise<Outcome> {
    if (this.active.has(request.id)) throw new Error("Duplicate execution id");
    const start = Date.now();
    let alive = true;
    const logs: ExecutionLog[] = [];
    const cleanups: Array<() => void> = [];
    const sleeps = new Map<() => void, (error: Error) => void>();
    const log = (level: ExecutionLog["level"], values: unknown[]) => {
      if (!alive || logs.length >= MAX_LOGS) return;
      const message = logText(values);
      logs.push({ level, message });
      this.runtime.progress?.({
        id: request.id,
        seq: logs.length,
        log: { level, message },
      });
      this.runtime.print(level, `[${request.id}] ${message}`);
    };
    let rejectCancelled: (error: Error) => void = () => {};
    const cancelled = new Promise<never>((_resolve, reject) => {
      rejectCancelled = reject;
    });
    const cancel = () => {
      if (!alive) return;
      alive = false;
      const error = new Error(
        "Execution cancelled or deadline exceeded; unmanaged effects may continue.",
      );
      for (const [stopTimer, reject] of sleeps) {
        stopTimer();
        reject(error);
      }
      sleeps.clear();
      rejectCancelled(error);
    };
    this.active.set(request.id, cancel);
    const stopDeadline = scheduleTimeout(cancel, request.timeoutMs);
    const ctx: Context = {
      id: request.id,
      resource: this.runtime.resource,
      alive: () => alive,
      game: <T>(callback: () => T) => {
        if (!alive)
          return Promise.reject(new Error("Execution is no longer active"));
        const guarded = () => {
          if (!alive) throw new Error("Execution is no longer active");
          return callback();
        };
        return this.runtime.game
          ? this.runtime.game(guarded)
          : Promise.resolve().then(guarded);
      },
      log: (...values) => log("info", values),
      sleep: (ms) => {
        if (!alive)
          return Promise.reject(new Error("Execution is no longer active"));
        if (!Number.isFinite(ms) || ms < 0 || ms > 60_000)
          return Promise.reject(
            new Error("sleep expects 0..60000 milliseconds"),
          );
        return new Promise((resolve, reject) => {
          const stopTimer = scheduleTimeout(() => {
            sleeps.delete(stopTimer);
            resolve();
          }, ms);
          sleeps.set(stopTimer, reject);
        });
      },
      onCleanup: (callback) => {
        if (!alive) throw new Error("Execution is no longer active");
        if (typeof callback !== "function")
          throw new Error("onCleanup expects a function");
        if (cleanups.length >= 100)
          throw new Error("Too many cleanup callbacks");
        cleanups.push(callback);
      },
    };
    const consoleProxy = {
      log: ctx.log,
      info: ctx.log,
      debug: (...values: unknown[]) => log("debug", values),
      warn: (...values: unknown[]) => log("warn", values),
      error: (...values: unknown[]) => log("error", values),
    };
    let outcome: Outcome;
    try {
      const snippet = new AsyncFunction(
        "ctx",
        "console",
        "require",
        "exports",
        `"use strict";\n${request.code}\n//# sourceURL=@${this.runtime.resource}/mcp/${request.id}.js`,
      );
      const value = await Promise.race([
        snippet(ctx, consoleProxy, this.runtime.require, this.runtime.exports),
        cancelled,
      ]);
      outcome = {
        ok: true,
        values: [sanitize(value)],
        logs,
        durationMs: Date.now() - start,
      };
    } catch (error) {
      outcome = {
        ok: false,
        values: [],
        error: errorMessage(error),
        logs,
        durationMs: Date.now() - start,
      };
    } finally {
      stopDeadline();
      this.active.delete(request.id);
      alive = false;
      for (const [stopTimer, reject] of sleeps) {
        stopTimer();
        reject(new Error("Execution finished"));
      }
      sleeps.clear();
    }
    for (const cleanup of cleanups.reverse()) {
      try {
        cleanup();
      } catch (error) {
        outcome.ok = false;
        outcome.error =
          `${outcome.error ?? ""}\nCleanup failed: ${errorMessage(error)}`.slice(
            0,
            8192,
          );
      }
    }
    return outcome;
  }
}

File: lua/executor.lua
local resource = GetCurrentResourceName()
local prefix = resource .. ':mcp:lua:'
local active = {}
local MAX_LOGS = 100
local MAX_RESULT = 131072

local function arrayJson(values, count)
    local parts = {}
    for i = 1, count or #values do
        parts[i] = json.encode(values[i])
    end
    return '[' .. table.concat(parts, ',') .. ']'
end

local function sanitize(value, depth, seen, budget)
    budget = budget or { remaining = 1000 }
    budget.remaining = budget.remaining - 1
    if budget.remaining < 0 then return { ['$type'] = 'max-items' } end
    local kind = type(value)
    if kind == 'nil' then return { ['$type'] = 'nil' } end
    if kind == 'string' then return #value > 8192 and value:sub(1, 8192) .. '[truncated]' or value end
    if kind == 'boolean' then return value end
    if kind == 'number' then
        if value ~= value or value == math.huge or value == -math.huge then
            return { ['$type'] = 'number', value = tostring(value) }
        end
        return value
    end
    if kind == 'vector2' or kind == 'vector3' or kind == 'vector4' then
        local vector = { ['$type'] = kind, x = value.x, y = value.y }
        if kind ~= 'vector2' then vector.z = value.z end
        if kind == 'vector4' then vector.w = value.w end
        return vector
    end
    if kind ~= 'table' then return { ['$type'] = kind } end
    if seen[value] then return { ['$type'] = 'circular' } end
    if depth >= 6 then return { ['$type'] = 'max-depth' } end
    seen[value] = true
    local result, entries, count, stringKeys = {}, {}, 0, true
    for key, item in next, value do
        count = count + 1
        if count > 100 then
            result['$truncated'] = true
            break
        end
        local safe = sanitize(item, depth + 1, seen, budget)
        if type(key) == 'string' then
            result[key:sub(1, 256)] = safe
        else
            stringKeys = false
        end
        entries[#entries + 1] = { key = sanitize(key, depth + 1, seen, budget), value = safe }
    end
    seen[value] = nil
    if stringKeys then return result end
    return { ['$type'] = 'table', entries = entries, truncated = count > 100 }
end

local function failure(id, message)
    TriggerEvent(prefix .. 'result', id, json.encode({
        ok = false, error = tostring(message):sub(1, 8192), durationMs = 0
    }):sub(1, -2) .. ',"values":[],"logs":[]}')
end

AddEventHandler(prefix .. 'execute', function(raw)
    if type(raw) ~= 'string' or #raw > 270000 then
        print('[dolu_fivem_mcp] Rejected invalid Lua execution envelope')
        return
    end
    local decoded, request = pcall(json.decode, raw)
    if not decoded or type(request) ~= 'table' or type(request.id) ~= 'string'
        or type(request.code) ~= 'string' or #request.code > 65536
        or type(request.timeoutMs) ~= 'number' or request.timeoutMs < 100 or request.timeoutMs > 60000 then
        print('[dolu_fivem_mcp] Rejected malformed Lua execution request')
        return
    end
    if active[request.id] then
        failure(request.id, 'Duplicate execution id')
        return
    end
    local state = { alive = true }
    active[request.id] = state
    CreateThread(function()
        local start = GetGameTimer()
        local logs, cleanups = {}, {}
        local function log(...)
            if not state.alive or #logs >= MAX_LOGS then return end
            local values = table.pack(...)
            local parts = {}
            for i = 1, values.n do
                parts[i] = type(values[i]) == 'string' and values[i] or json.encode(sanitize(values[i], 0, {}))
            end
            local message = table.concat(parts, ' '):sub(1, 4096)
            logs[#logs + 1] = { level = 'info', message = message }
            TriggerEvent(prefix .. 'log', json.encode({
                id = request.id, seq = #logs, log = logs[#logs]
            }))
            print(('[%s][%s] %s'):format(resource, request.id, message))
        end
        local ctx = {
            id = request.id,
            resource = resource,
            alive = function() return state.alive and GetGameTimer() - start < request.timeoutMs end,
            log = log,
            sleep = function(ms)
                assert(type(ms) == 'number' and ms >= 0 and ms <= 60000, 'sleep expects 0..60000 milliseconds')
                assert(state.alive, 'Execution cancelled')
                Wait(ms)
                assert(state.alive and GetGameTimer() - start < request.timeoutMs, 'Execution cancelled or deadline exceeded')
            end,
            onCleanup = function(callback)
                assert(state.alive, 'Execution is no longer active')
                assert(type(callback) == 'function' and #cleanups < 100, 'Invalid or excessive cleanup callback')
                cleanups[#cleanups + 1] = callback
            end
        }
        local env = setmetatable({ ctx = ctx, print = log }, { __index = _G })
        local chunk, compileError = load(request.code, '@' .. resource .. '/mcp/' .. request.id .. '.lua', 't', env)
        if not chunk then
            active[request.id] = nil
            state.alive = false
            failure(request.id, compileError)
            return
        end
        -- Best-effort guard, not a sandbox: trusted code can bypass hooks or block in a native.
        local guarded = type(debug) == 'table' and type(debug.sethook) == 'function'
        local instructions = 0
        if guarded then
            debug.sethook(function()
                instructions = instructions + 10000
                if instructions >= 10000000 or not ctx.alive() then
                    error('Lua instruction budget or execution deadline exceeded', 0)
                end
            end, '', 10000)
        end
        local returned = table.pack(xpcall(function() return chunk(ctx) end, function(err)
            return debug and debug.traceback and debug.traceback(tostring(err), 2) or tostring(err)
        end))
        if guarded then debug.sethook() end
        state.alive = false
        active[request.id] = nil
        local ok = returned[1]
        local err
        if not ok then err = tostring(returned[2]) end
        for i = #cleanups, 1, -1 do
            local cleaned, cleanupError = pcall(cleanups[i])
            if not cleaned then
                ok = false
                err = (err or '') .. '\nCleanup failed: ' .. tostring(cleanupError)
            end
        end
        local encoded, result = pcall(function()
            local values = {}
            if returned[1] then
                for i = 2, math.min(returned.n, 101) do
                    values[#values + 1] = sanitize(returned[i], 0, {})
                end
            end
            return json.encode({
                ok = ok, error = err and err:sub(1, 8192),
                durationMs = math.max(0, GetGameTimer() - start),
                truncated = returned.n > 101
            }):sub(1, -2) .. ',"values":' .. arrayJson(values) .. ',"logs":' .. arrayJson(logs) .. '}'
        end)
        if not encoded then
            failure(request.id, 'Result serialization failed: ' .. tostring(result))
        elseif #result > MAX_RESULT then
            failure(request.id, 'Result exceeds transport size limit; return a smaller value')
        else
            TriggerEvent(prefix .. 'result', request.id, result)
        end
    end)
end)

AddEventHandler(prefix .. 'cancel', function(id)
    if active[id] then active[id].alive = false end
end)

local function stopping(name)
    if name ~= resource then return end
    for _, state in pairs(active) do state.alive = false end
end
AddEventHandler('onResourceStop', stopping)
AddEventHandler('onClientResourceStop', stopping)
